webofneeds/won-owner/src/main/java/won/owner/service/impl/KeystorePasswordUtils.java

```diff
@@ -0,0 +1,151 @@
+package won.owner.service.impl;
+
+import java.math.BigInteger;
+import java.security.NoSuchAlgorithmException;
+import java.security.SecureRandom;
+import java.security.spec.InvalidKeySpecException;
+import java.util.Arrays;
+import java.util.Base64;
+
+import javax.crypto.Cipher;
+import javax.crypto.SecretKeyFactory;
+import javax.crypto.spec.IvParameterSpec;
+import javax.crypto.spec.PBEKeySpec;
+import javax.crypto.spec.SecretKeySpec;
+import javax.xml.bind.DatatypeConverter;
+
+import org.springframework.util.StopWatch;
+
+public class KeystorePasswordUtils {
+	public static final String CURRENT_VERSION = "v1";
+	public static final int KEYSTORE_PASSWORD_BYTES = 32;
+
+	/*
+	public static void main(String... args) {
+		String password = "supersecretpassword";
+		String key = "mykey";
+		StopWatch sw = new StopWatch();
+		sw.start();
+		System.out.println("encrypting...");
+		String encrypted = encryptPassword(password, key);
+		sw.stop();
+		System.out.println("took " + sw.getLastTaskTimeMillis());
+		sw.start();
+		System.out.println("encrypting " + password + " with key " + key + " to: " + encrypted);
+		String decrypted = decryptPassword(encrypted, key);
+		System.out.println("decrypting again yields: " + decrypted);
+		sw.stop();
+		System.out.println("took " + sw.getLastTaskTimeMillis());
+
+		System.out.println("generating password: " + generatePassword("abebu", 256, null));
+	}*/
+
+	public static String encryptPassword(String password, String key) {
+		try {
+			int iterations = 1000;
+			byte[] salt = getSalt();
+			byte[] iv = getSalt();
+			PBEKeySpec spec = new PBEKeySpec(key.toCharArray(), salt, iterations, 256);
+			SecretKeyFactory skf = SecretKeyFactory.getInstance("PBKDF2WithHmacSHA1");
+			byte[] encryptionKey = skf.generateSecret(spec).getEncoded();
+			SecretKeySpec secretKey = new SecretKeySpec(encryptionKey, "AES");
+			Cipher cipher = Cipher.getInstance(secretKey.getAlgorithm() + "/CFB8/NoPadding");
+			cipher.init(Cipher.ENCRYPT_MODE, secretKey, new IvParameterSpec(iv));
+			return CURRENT_VERSION + ":" + iterations + ":" + toHex(salt) + ":" + toHex(iv) + ":"
+					+ toHex(cipher.doFinal(password.getBytes()));
+		} catch (Exception e) {
+			throw new IllegalArgumentException("cannot encrypt password", e);

```

Comment:
maybe we should avoid using the password in a exception that gets logged/printed

Reply:
You mean, because we cannot know what e.getMessage() will say? I guess that's a valid point - we could just report the class of the exception in our new exception's message, to give some hint at what went wrong.

Reply:
oh no, sorry i misread the code, i thought the password was printed

Reply:
all fine